Add 'make docker' for a one-line Docker build

samstokes opened the pull request:
It's convenient to build inside Docker so as not to pollute your local laptop with development dependencies.  However, the build process has several steps (build it, extract the tarballs, build the individual images).  This wraps them up into a single command, `make docker`.

This solution has a few gotchas:

 * I'm using `docker run cat` to extract the tarballs, which doesn't preserve timestamps, so one or both of Docker and `make` detect it being a fresh build every time, even if you didn't change any source files.  This might be mitigated by changing Dockerfile.build (I suspect the `COPY .` line is the culprit, since it picks up the tarballs too) or dumping the tarballs in /tmp instead of the working directory.

 * This also means each time you build it creates a new Docker image, rather than reusing very much from the image cache, so you can fill up your disk over a stretch.

However, I think it's a step forward.

ept commented:
@samstokes Yes, I think this is a big improvement, thank you.

However, on my Mac OS, make does not recognise the `*.tar.gz` rule (using GNU Make 3.81). The make manual suggests that a [pattern rule](http://www.gnu.org/software/make/manual/make.html#Pattern-Rules) should use `%` rather than `*` as wildcard character, and if I change the target to `%.tar.gz` it works. Does that work for you too?

Another suggestion: to avoid cluttering the root directory of the repository, `make docker` could put the `.tar.gz` files in another directory (e.g. `build/`). Also, it would be good to add them to `.gitignore`.

Finally, would you mind updating the docs at the top of `Dockerfile.build` to match the new build process?

samstokes commented:
Thanks for the review!  I've addressed your comments in the commits above.

What do you think about 31a32cd?  (Builds the client and postgres images from within the tmp directory instead of the project root - more info in the commit message.)  It improves the standard dev lifecycle (`make docker` now caches better, so runs faster and fills up your disk less), but makes it slightly more confusing for anyone consuming those Dockerfiles directly (i.e. not via `make docker`).  I've documented the new behaviour in the Dockerfiles, but wanted to get a second opinion.

ept commented:
Sorry for the delayed review. This looks great, thank you!
